src/lib/datatypes/sol-str-slice.c

```diff
@@ -48,16 +48,18 @@ sol_str_slice_to_int(const struct sol_str_slice s, int *value)
     return 0;
 }
 
-SOL_API bool
+SOL_API char *
 sol_str_slice_contains(const struct sol_str_slice haystack, const struct sol_str_slice needle)
 {
-    return memmem(haystack.data, haystack.len, needle.data, needle.len) != NULL;
+    return memmem(haystack.data, haystack.len, needle.data, needle.len);
 }
 
-SOL_API bool
+SOL_API char *
 sol_str_slice_str_contains(const struct sol_str_slice haystack, const char *needle)
 {
-    return memmem(haystack.data, haystack.len, needle, strlen(needle)) != NULL;
+    SOL_NULL_CHECK(needle, NULL);
+
+    return memmem(haystack.data, haystack.len, needle, strlen(needle));
```

Comment:
why not add these to header and convert to slice? Then all the slice conversions are done in header, where compiler will see `strlen()` and can optimize constants.